api/src/main/java/io/minio/MinioClient.java

```diff
@@ -1397,6 +1399,217 @@ public void getObject(String bucketName, String objectName, String fileName)
+   *          CopyConditions object with collection of supported CopyObject conditions.
+   *
+   * @throws InvalidBucketNameException
+   *           upon an invalid bucket name
+   * @throws NoSuchAlgorithmException
+   *           upon requested algorithm was not found during signature calculation
+   * @throws InvalidKeyException
+   *           upon an invalid access key or secret key
+   */
+  public void copyObject(String bucketName, String objectName, String destBucketName,
+                         String destObjectName, CopyConditions copyConditions)
+      throws InvalidKeyException, InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException,
+      NoResponseException, ErrorResponseException, InternalException, IOException, XmlPullParserException,
+      InvalidArgumentException {
+
+    if (destObjectName == null) {
+      destObjectName = objectName;
+    }
+
+    String sourceObjectPath = S3Escaper.encode(Paths.get(bucketName, objectName).toString());
+    copyObjectResult(destBucketName, destObjectName, sourceObjectPath, copyConditions);
+  }
+
+  /**
+   * Executes CopyObject and returns CopyObjectResult of the CopyObject operation.
+   *
+   * @param destBucketName
+   *          Bucket name where the object will be created.
+   * @param destObjectName
+   *          Object name to be created after copy.
+   * @param sourceObjectPath
+   *          Path to the source object to be copied, is always
+   *          prefixed with a bucket.
+   * @param copyConditions
+   *          CopyConditions object with collection of supported CopyObject conditions.
+   *
+   * @throws InvalidBucketNameException
+   *           upon an invalid bucket name
+   * @throws NoSuchAlgorithmException
+   *           upon requested algorithm was not found during signature calculation
+   * @throws InvalidKeyException
+   *           upon an invalid access key or secret key
+   */
+  private CopyObjectResult copyObjectResult(String destBucketName, String destObjectName, String sourceObjectPath,
+                                            CopyConditions copyConditions)
+      throws InvalidKeyException, InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException,
+      NoResponseException, ErrorResponseException, InternalException, IOException, XmlPullParserException,
+      InvalidArgumentException {
+
+    if (sourceObjectPath == null) {
+      throw new InvalidArgumentException("Object source cannot be empty");
+    }
+
+    Map<String, String> headerMap = new HashMap<>();
+
+    // Set the object source
+    headerMap.put("x-amz-copy-source", sourceObjectPath);
+
+    // If no conditions available, skip addition else add the conditions to the header
+    if (copyConditions != null) {
+      headerMap.putAll(copyConditions.getConditions());
+    }
+
+    updateRegionCache(destBucketName);
+    HttpResponse response =  execute(Method.PUT, BucketRegionCache.INSTANCE.region(destBucketName),
```

Comment:
My suggestion is to fix existing `executePut()` returning unclosed body and use it here.